File: Ports/jdk/jdk-jdk-18-10/src/jdk.jpackage/windows/native/msiwrapper/Executor.cpp
/*
 * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#include <algorithm>
#include "Executor.h"
#include "Log.h"
#include "WinErrorHandling.h"


namespace {

void escapeArg(std::wstring& str) {
    if (str.empty()) {
        return;
    }

    if (str.front() == L'\"' && str.back() == L'\"' && str.size() > 1) {
        return;
    }

    if (str.find_first_of(L" \t") != std::wstring::npos) {
        str = L'"' + str + L'"';
    }
}

} // namespace


std::wstring Executor::args() const {
    tstring_array tmpArgs;
    // argv[0] is the module name.
    tmpArgs.push_back(appPath);
    tmpArgs.insert(tmpArgs.end(), argsArray.begin(), argsArray.end());

    std::for_each(tmpArgs.begin(), tmpArgs.end(), escapeArg);
    return tstrings::join(tmpArgs.begin(), tmpArgs.end(), _T(" "));
}


int Executor::execAndWaitForExit() const {
    UniqueHandle h = startProcess();

    const DWORD res = ::WaitForSingleObject(h.get(), INFINITE);
    if (WAIT_FAILED ==  res) {
        JP_THROW(SysError("WaitForSingleObject() failed", WaitForSingleObject));
    }

    DWORD exitCode = 0;
    if (!GetExitCodeProcess(h.get(), &exitCode)) {
        // Error reading process's exit code.
        JP_THROW(SysError("GetExitCodeProcess() failed", GetExitCodeProcess));
    }

    const DWORD processId = GetProcessId(h.get());
    if (!processId) {
        JP_THROW(SysError("GetProcessId() failed.", GetProcessId));
    }

    LOG_TRACE(tstrings::any() << "Process with PID=" << processId
                                << " terminated. Exit code=" << exitCode);

    return static_cast<int>(exitCode);
}


UniqueHandle Executor::startProcess() const {
    const std::wstring argsStr = args();

    std::vector<TCHAR> argsBuffer(argsStr.begin(), argsStr.end());
    argsBuffer.push_back(0); // terminating '\0'

    STARTUPINFO startupInfo;
    ZeroMemory(&startupInfo, sizeof(startupInfo));
    startupInfo.cb = sizeof(startupInfo);

    PROCESS_INFORMATION processInfo;
    ZeroMemory(&processInfo, sizeof(processInfo));

    DWORD creationFlags = 0;

    if (!theVisible) {
        // For GUI applications.
        startupInfo.dwFlags |= STARTF_USESHOWWINDOW;
        startupInfo.wShowWindow = SW_HIDE;

        // For console applications.
        creationFlags |= CREATE_NO_WINDOW;
    }

    tstrings::any msg;
    msg << "CreateProcess(" << appPath << ", " << argsStr << ")";

    if (!CreateProcess(appPath.c_str(), argsBuffer.data(), NULL, NULL, FALSE,
                    creationFlags, NULL, NULL, &startupInfo, &processInfo)) {
        msg << " failed";
        JP_THROW(SysError(msg, CreateProcess));
    }

    msg << " succeeded; PID=" << processInfo.dwProcessId;
    LOG_TRACE(msg);

    // Close unneeded handles immediately.
    UniqueHandle(processInfo.hThread);

    // Return process handle.
    return UniqueHandle(processInfo.hProcess);
}

File: Ports/jdk/jdk-jdk-18-10/test/hotspot/jtreg/compiler/compilercontrol/share/scenario/Executor.java
/*
 * Copyright (c) 2015, 2017, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package compiler.compilercontrol.share.scenario;

import compiler.compilercontrol.share.actions.BaseAction;
import jdk.test.lib.Asserts;
import jdk.test.lib.management.InputArguments;
import jdk.test.lib.process.OutputAnalyzer;
import jdk.test.lib.process.ProcessTools;
import jdk.test.lib.dcmd.CommandExecutor;
import jdk.test.lib.dcmd.PidJcmdExecutor;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.lang.reflect.Executable;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class Executor {
    private final boolean isValid;
    private final List<String> vmOptions;
    private final Map<Executable, State> states;
    private final List<String> jcmdCommands;
    private final String execClass = System.getProperty("compiler."
            + "compilercontrol.share.executor.executeClass",
            BaseAction.class.getName());
    private OutputAnalyzer[] jcmdOutputAnalyzers;

    /**
     * Constructor
     *
     * @param isValid      shows that the input given to the VM is valid and
     *                     VM shouldn't fail
     * @param vmOptions    a list of VM input options
     * @param states       a state map, or null for the non-checking execution
     * @param jcmdCommands a list of diagnostic commands to be preformed
     *                     on test VM
     */
    public Executor(boolean isValid, List<String> vmOptions,
            Map<Executable, State> states, List<String> jcmdCommands) {
        this.isValid = isValid;
        if (vmOptions == null) {
            this.vmOptions = new ArrayList<>();
        } else {
            this.vmOptions = vmOptions;
        }
        this.states = states;
        this.jcmdCommands = jcmdCommands;
    }

    /**
     * Executes separate VM a gets an OutputAnalyzer instance with the results
     * of execution
     */
    public List<OutputAnalyzer> execute() {
        // Add class name that would be executed in a separate VM
        vmOptions.add(execClass);
        OutputAnalyzer output;
        try (ServerSocket serverSocket = new ServerSocket(0)) {
            {
                // Get port test VM will connect to
                int port = serverSocket.getLocalPort();
                if (port == -1) {
                    throw new Error("Socket is not bound: " + port);
                }
                vmOptions.add(String.valueOf(port));
                if (states != null) {
                    // add flag to show that there should be state map passed
                    vmOptions.add("states");
                }
                // Start separate thread to connect with test VM
                new Thread(() -> connectTestVM(serverSocket)).start();
            }
            // Start a test VM using vm flags from @run and from vm options
            String[] vmInputArgs = InputArguments.getVmInputArgs();
            String[] cmds = Arrays.copyOf(vmInputArgs,
                    vmInputArgs.length + vmOptions.size());
            System.arraycopy(vmOptions.toArray(), 0, cmds, vmInputArgs.length,
                    vmOptions.size());
            output = ProcessTools.executeTestJvm(cmds);
        } catch (Throwable thr) {
            throw new Error("Execution failed: " + thr.getMessage(), thr);
        }

        List<OutputAnalyzer> outputList = new ArrayList<>();
        outputList.add(output);
        if (jcmdOutputAnalyzers != null) {
            Collections.addAll(outputList, jcmdOutputAnalyzers);
        }
        return outputList;
    }

    /*
     * Performs connection with a test VM, sends method states and performs
     * JCMD operations on a test VM.
     */
    private void connectTestVM(ServerSocket serverSocket) {
        /*
         * There are no way to prove that accept was invoked before we started
         * test VM that connects to this serverSocket. Connection timeout is
         * enough
         */
        try (
                Socket socket = serverSocket.accept();
                PrintWriter pw = new PrintWriter(socket.getOutputStream(),
                        true);
                BufferedReader in = new BufferedReader(new InputStreamReader(
                        socket.getInputStream()))) {
            // Get pid of the executed process
            int pid = Integer.parseInt(in.readLine());
            Asserts.assertNE(pid, 0, "Got incorrect pid");
            jcmdOutputAnalyzers = executeJCMD(pid);
            if (states != null) {
                // serialize and send state map
                states.forEach((executable, state) -> {
                    pw.println("{");
                    pw.println(executable.toGenericString());
                    pw.println(state.toString());
                    pw.println("}");
                });
            } else {
                pw.println();
            }
        } catch (IOException e) {
            // hotspot process may exit because of invalid directives,
            // suppress the IOException of closed socket.
            if (!e.getMessage().equals("Socket closed")) {
              throw new Error("Failed to write data: " + e.getMessage(), e);
            }
        }
    }

    // Executes all diagnostic commands
    protected OutputAnalyzer[] executeJCMD(int pid) {
        int size = jcmdCommands.size();
        OutputAnalyzer[] outputArray = new OutputAnalyzer[size];
        CommandExecutor jcmdExecutor = new PidJcmdExecutor(String.valueOf(pid));
        for (int i = 0; i < size; i++) {
            outputArray[i] = jcmdExecutor.execute(jcmdCommands.get(i));
        }
        return outputArray;
    }
}
